### PR TITLE
Feature: Refactor markets

### PR DESCRIPTION
**Pull request by VladyslavP**

### CODE REVIEW

File: src/components/Markets/Markets.test.tsx [hunk 1]
```diff
@@ -34,26 +36,26 @@ describe('Markets', () => {
...
-          ['ETH/BTC', '0.123', '+50.00%'],
-      ];
-
-      (component.instance() as Markets).handleFilter(filteredData as object[]);
-      const input = component.find('tr').first();
-      input.simulate('click');
-
-      expect((component.state() as MarketsState).filteredData).toEqual(filteredData);
-   });
+   // it('should correctly filter rows', () => {
```
Comment (oyershov) [inline src/components/Markets/Markets.test.tsx:39]: please, fix test to merge this PR

File: src/components/Markets/index.tsx [hunk 2]
```diff
@@ -123,52 +65,85 @@ export class Markets extends React.Component<MarketsProps, MarketsState> {
...
-        filter: this.filterType('Pair', currency),
-    });
-}
+    const getMarketFromDataRow = (market: React.ReactNode[]) => market[0] as string;
+
+    const { filters = true, headers, title, filterPlaceholder = '', rowKeyIndex, selectedKey } = props;
+
+    let tableData = filteredData.length > 0
+        ? filteredData
+        : [['', '', '']];
```
Comment (oyershov) [inline src/components/Markets/index.tsx:121]: Need to change ```[['', '', '']]``` and pass just empty array ```[]``` and change all places like this

File: src/components/Markets/index.tsx [hunk 3]
```diff
@@ -123,52 +65,85 @@ export class Markets extends React.Component<MarketsProps, MarketsState> {
...
+
+    tableData = tableData.map(row => row.map(mapRows));
+
+    // TODO: Refactor this logic
+    React.useEffect(() => {
+        setFilteredData(props.data.filter(w => (w[0] as string).toLowerCase().includes(searchKey.toLowerCase())));
+    }, [props.data, searchKey]);
```
Comment (oyershov) [inline src/components/Markets/index.tsx:128]: please, refactor in this PR

File: src/components/Markets/index.tsx [hunk 4]
```diff
@@ -123,52 +65,85 @@ export class Markets extends React.Component<MarketsProps, MarketsState> {
...
     };
 
-    private get filters() {
-        const { data } = this.props;
+    const transformCurrencyToFilter = (currency: string) => ({
+        name: currency,
+        filter: filterType('Pair', currency),
+    });
+
+    const filtersFn = () => {
```
Comment (oyershov) [inline src/components/Markets/index.tsx:96]: try to find better name for this function

File: src/containers/Markets/index.tsx [hunk 5]
```diff
@@ -18,117 +15,83 @@ import {
...
+        'pg-markets--loading': marketsLoading,
     });
 
-export const MarketsComponent = injectIntl(connect(mapStateToProps, mapDispatchToProps)(MarketsContainer)) as any;
+
+    React.useEffect(() => {
+        if (marketsData.length === 0) {
+            dispatch(marketsTickersFetch());
+        }
+    }, [marketsData.length, dispatch]);
```
Comment (oyershov) [inline src/containers/Markets/index.tsx:90]: use hook ```useMarketsTickersFetch()```

File: src/containers/Markets/index.tsx [hunk 6]
```diff
@@ -18,117 +15,83 @@ import {
...
+    const handleOnSelect = (index: string) => {
+        const marketToSet = marketsData.find(el => el.name === index);
+        dispatch(setCurrentPrice(0));
 
         if (marketToSet && (!currentMarket || currentMarket.id !== marketToSet.id)) {
-            this.props.setCurrentMarket(marketToSet);
+            dispatch(setCurrentMarket(marketToSet));
             if (!incrementalOrderBook()) {
-              this.props.depthFetch(marketToSet);
+                dispatch(depthFetch(marketToSet));
```
Comment (oyershov) [inline src/containers/Markets/index.tsx:76]: use hook for this ```useDepthFetch()```

File: src/components/Markets/index.tsx [hunk 7]
```diff
@@ -123,52 +65,85 @@ export class Markets extends React.Component<MarketsProps, MarketsState> {
...
 
         const currencyFilters = data && data.length > 0
-            ? this.props.data
-                .map(this.getMarketFromDataRow)
-                .reduce(this.createUniqueCurrencies, [])
-                .map(this.transformCurrencyToFilter)
+            ? data
+                .map(getMarketFromDataRow)
+                .reduce(createUniqueCurrencies, [])
+                .map(transformCurrencyToFilter)
```
Comment (oyershov) [inline src/components/Markets/index.tsx:103]: I'm sure this needs to be refactored.

**Review by oyershov — CHANGES_REQUESTED** [review]

File: src/components/Markets/Markets.test.tsx (hunk 1 above)
Comment (VladyslavP) [reply src/components/Markets/Markets.test.tsx:39]: In order to fix tests we have to support tests for hooks. I suggest to do it as a part of another task

File: src/components/Markets/index.tsx (hunk 2 above)
Comment (VladyslavP) [reply src/components/Markets/index.tsx:121]: Not quite sure what you mean. It was like this before my changes

File: src/components/Markets/index.tsx (hunk 4 above)
Comment (VladyslavP) [reply src/components/Markets/index.tsx:96]: Previously, it was a getter for filters, right now it is a function

File: src/helpers/isUniqueValue.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const isUniqueValue = (arr, value) => !arr.includes(value);
```
Comment (oyershov) [inline src/helpers/isUniqueValue.ts:1]: 1. Define types
2. Add check if object exists
3. Cover helper by tests

File: src/containers/Markets/index.tsx [hunk 9]
```diff
@@ -18,117 +15,83 @@ import {
...
-        return (
-            <div className={className}>
-                {marketsLoading ? <div><Spinner animation="border" variant="primary" /></div> : this.markets()}
-            </div>
-        );
-    }
 
-    private markets = () => {
-        const { currentMarket } = this.props;
+    const markets = () => {
```
Comment (oyershov) [inline src/containers/Markets/index.tsx:37]:
```suggestion
    const renderMarkets = () => {
```

File: src/containers/Markets/index.tsx [hunk 10]
```diff
@@ -18,117 +15,83 @@ import {
...
         );
-    }
+    };
 
-    private handleOnSelect = (index: string) => {
-        const { markets, currentMarket } = this.props;
-        const marketToSet = markets.find(el => el.name === index);
-        this.props.setCurrentPrice(0);
+    const handleOnSelect = (index: string) => {
+        const marketToSet = marketsData.find(el => el.name === index);
```
Comment (oyershov) [inline src/containers/Markets/index.tsx:70]: marketsData can be undefined, please add check

File: src/containers/Markets/index.tsx (hunk 5 above)
Comment (oyershov) [inline src/containers/Markets/index.tsx:90]: use ```useMarketsTickersFetch()``` hook

**Review by oyershov — CHANGES_REQUESTED** [review]

File: src/components/Markets/index.tsx [hunk 11]
```diff
@@ -123,52 +60,75 @@ export class Markets extends React.Component<MarketsProps, MarketsState> {
...
     };
 
-    private get filters() {
-        const { data } = this.props;
+    const transformCurrencyToFilter = (currency: string) => ({
+        name: currency,
+        filter: filterType('Pair', currency),
+    });
+
+    const getFilters = () => {
```
Comment (berezh) [inline src/components/Markets/index.tsx:89]: use `useCallback` hook for ALL inner methods inside component

**Review by oyershov — APPROVED** [review]

**Review by berezh — APPROVED** [review]

**Pull request merged**